### PR TITLE
changefeed: support primary key columns which are not visible

### LINKED ISSUE #128457: sql/schemachanger: ALTER PRIMARY KEY .. USING HASH can make the hash column public too late
Previously, in the legacy schema changer, when executing ALTER TABLE PRIMARY KEY .. USING HASH... we have stages where a new primary index is public without all the columns inside it being public. This is unexpected, since the key columns / storing referenced in the primary key should be public at the same time as the index. This can cause issues for CDC: https://github.com/wenyihu6/cockroach/blob/2d4f2a14a05b8e23a134cad42b435886657f8172/pkg/ccl/changefeedccl/cdcevent/event.go#L312

We should ideally make sure the columns referenced by the primary key index are public at the same time. Or at the very least ensure this is guaranteed for key columns.

Jira issue: CRDB-41028

Epic CRDB-40419

### PR DESCRIPTION
**Pull request by fqazi**
Previously, when schema changes were happening with an active change feed, it was possible for ALTER PRIMARY KEY USING HASH to have an intermediate state where the hash column could exist within the primary key without being public. This is expected since complex schema changes could not roll back if these columns were made completely public at the same time as the primary key. This would break assumptions within CDC, which expected all primary key columns to be public. To address this, this patch modifies this logic to ignore columns that are writable but not yet fully readable in the primary key.

Fixes: #128457

Release note: None

### CODE REVIEW

**Review by asg0451 — COMMENTED** [review]
sorry for the delay, i'm having to learn what all of this means as i go.

File: pkg/ccl/changefeedccl/cdcevent/event_test.go [hunk 1]
```diff
@@ -134,6 +139,147 @@ CREATE TABLE foo (
...
+		{
+			// Replace the primary index with a new one with a hash. This will lead
+			// to 3 indexes observed:
+			// 1) The original from the previous test
+			// 2) An index without the hash and just c + e
+			// 3) An index with c + e and a new hash.
+			schemaChange:    "ALTER TABLE foo ALTER PRIMARY KEY USING COLUMNS(c, e) USING HASH",
+			expectedKeyCols: [][]string{{"crdb_internal_a_b_shard_16", "b", "a"}, {"c", "e"}, {"crdb_internal_c_e_shard_16", "c", "e"}},
+			expectedColumns: [][]string{{"a", "b", "e"}, {"a", "b", "e"}, {"a", "b", "e"}},
+			expectedUDTCols: [][]string{{"e"}, {"e"}, {"e"}},
```
Comment (asg0451) [inline pkg/ccl/changefeedccl/cdcevent/event_test.go:201]: I'm having a hard time understanding why these columns are the way they are - why is `{"a", "b", "e"}` repeated 3 times? what's with the internal_shard thing? can you add some comments to explain this a little bit please?

**Review by asg0451 — COMMENTED** [review]
the code looks reasonable but i'm finding the tests inscrutable. can you add some explaining comments please?

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129828)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451 and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/cdcevent/event_test.go` line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129828#gh-1765393448:-O7ijI3n6fwUyYhYw4A-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/631a3f73a882b1435567ba46c30729d803ee6df7/pkg/ccl/changefeedccl/cdcevent/event_test.go#L143)):*
<details><summary><i>Previously, asg0451 (Miles Frankel) wrote…</i></summary><blockquote>

```suggestion
// will produce valid event descriptors.
```
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/cdcevent/event_test.go` line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129828#gh-1765468670:-O7jOSRZ3MPHoH3MyCMr:b-zgx1ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/631a3f73a882b1435567ba46c30729d803ee6df7/pkg/ccl/changefeedccl/cdcevent/event_test.go#L201)):*
<details><summary><i>Previously, asg0451 (Miles Frankel) wrote…</i></summary><blockquote>

I'm having a hard time understanding why these columns are the way they are - why is `{"a", "b", "e"}` repeated 3 times? what's with the internal_shard thing? can you add some comments to explain this a little bit please?
</blockquote></details>

I added a few comments here to clarify things:

```		
		// Each new primary index generated during the test will pause at each stage
		// of the schema change. Then it will try to see if the current index matches
		// a expected state, if it doesn't it will validate the against the next expected
		// state assuming a transition has occurred (i.e. the primary index has been
		// replaced).
```

Also clarified what steps are in each schema change, but the internal_shard column is how hash sharded indexes are implemented. i.e. For the first schema change:

```
			// Replace the primary index with one that uses hash sharding, we will see
			// the following states:
			// 1) the original primary key
			// 2) the new primary key, since the new primary key is hash sharded,
			//		we are going to have a new key column (crdb_internal_a_b_shard_16)
			//		that is hashing the primary key columns.
```

**Review by asg0451 — COMMENTED** [review]
Thanks for adding the comments, lgtm

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129828#-:-O8gTrZV9a2lpdluypc7:b4q9gch)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

**Review by asg0451 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129828#-:-O8gTx7S8Yhtu8MpKK5E:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)

**Comment by fqazi** [comment]
@asg0451 TFTR!

bors r+

**Pull request merged**